generator/internal/api/visitors.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import "fmt"
+
+// Element is an interface for all visitable API elements.
+type Element interface {
+	Accept(v *Visitor) error
+}
+
+// Visitor is an interface for visiting API elements, with each visitable element having its own Visit* method.
+//
+// Each node is visited before its children. Parent/child hierarchy is documented in api.API.
+//
+// Traversing the API with a visitor, even the NoOpVisitor, has the side effect of assigning the *.Parent and *.API
+// fields to the visited elements. This side effect is idempotent.
```

Comment:
That is too surprising for a  NoOpVisitor.

Reply:
We could (should?) move this logic to the open api and protobuf parsers. It just felt like a harmless shortcut to have it implemented in the visitor first, since it is idempotent and required only by visitors. If we ever need to traverse mixins *before* local types, we will need to change this, as the only way to distinguish them today is to traverse the local types first.

Reply:
The side effect itself isn't done by the visitor, it is done by the `Accept` methods in the model. Regardless of which visitor you provide, the first time you invoke `API.Accept`, all parent hierarchy will be set. 

The correct implementation here is for the `makeAPIFor*` functions to build the hierarchy correctly, including all the pointers (minus the cross-references). I can fix that as part of this PR to make sure the visitor implementation doesn't have any unintended side effect.